### ISSUE TITLE
Improve support for partial data sets

### ISSUE THREAD

**Issue by eneroth** [lilactown/pyramid#20]
I find myself sometimes working with data that is not entirely complete. For example:

```Clojure
;; OK

(def db (pyramid/db [{:id   0
                      :refs [[:id 1]]}
                     {:id    1
                      :hello :world}]))

(pyramid/pull db [{[:id 0]
                   [{:refs [:id :hello]}]}])

;; => {[:id 0] {:refs [{:id 1, :hello :world}]}}
```

I.e., I have a number of entities (for example, retrieved from persistent storage), and I'd like to throw them at Pyramid and ask some questions, ad-hoc. This works well, as long as the data set is complete (no dangling idents).

In the following example, it doesn't work as well:

```Clojure
;; Not OK

(def db (pyramid/db [{:id   0
                      :refs [[:id 1] [:id 2]]}
                     {:id    1
                      :hello :world}]))

(pyramid/pull db [{[:id 0]
                   [{:refs [:id :hello]}]}])

;; => Execution error (NullPointerException) at pyramid.pull/visit$fn (pull.cljc:157).
```

When working with Pathom in a similar scenario, the ident itself will be treated as a minimal entity (i.e., just converted to a map: `{:id 2}`), with the result of the pull (which may be `nil`) merged to that map. This guarantees that a minimal result will be returned where data is missing.

The result will be something like:

```Clojure
{[:id 0] {:refs [{:id 1, :hello :world}
                 {:id 2}]}}
```

What do you think of this?

**Comment by lilactown** [comment]
Thanks for the report!

This will work if you tweak the second example a bit - instead of passing in vectors as references, if you change them to maps, it will work:

```clojure
(def db (p/db [{:id   0
                :refs [{:id 1} {:id 2}]}
                {:id    1
                 :hello :world}]))

(p/pull db [{[:id 0]
             [{:refs [:id :hello]}]}])
;; => {[:id 0] {:refs [{:id 1, :hello :world} {:id 2}]}}
```

In general you can think of the operations like: `add` takes entities and returns data full of references, and `pull` takes data full of references and resolves entities. In the original, you're short circuiting the entity processing that `add` does.

My next thought is that when an entity doesn't exist in the db, to return _something_, which I guess gets to your ask- should a ref to an entity that doesn't exist return a minimal map `{:id 2}`, or nothing?

For my own uses, I think that eliding `[:id 2]` from the result in the original would be better. It means that if I dissoc an entity from the db, that dangling references aren't going to return incomplete data that I'll have to code defensively against. What do you think, in the context of the above solution (passing in maps for the refs when adding)?

Also, thank you for all the work you're doing to report these corners, @eneroth ! I really appreciate it. I hope that we can fine tune pyramid to work better for you.

**Comment by eneroth** [comment]
Looking at my original entry a bit more closely, I guess there are two levels to this:

1. Regardless, maybe an exception shouldn't be thrown here, at least?
2. If it's a dangling ident, should it be considered "not there"?

With regards to 2, my feeling is that by passing it in, I have effectively stated "there's something here, but I don't know much about it." Yeah, maybe the correct idea is that it should be turned into into a map then (I was hoping to avoid this, since it requires walking the entire structure, looking for idents, which I know that Pyramid will have to do _again_—feels wasteful).

With regards to eliding the value: my thinking here is formed from having worked with Fulcro, in which you're very much encouraged (primarily through helpers) to remove the entity _and_ it's reference. I.e., the DB doesn't contain anything that's open to interpretation, or could be considered "broken". In Fulcro, just dissocing an entity without cleaning up its references would be considered kind of sloppy, I think :) 

To be clear @lilactown, Pyramid already works great! I'm so happy to have something framework-agnostic that provides this functionality, and can be used more ad-hoc than a fullblown DB (even an in-memory one).

**Comment by lilactown** [comment]
yeah, there is a `pyramid.core/delete` fn which should handle it for you.

Fixed this in 94d2131e2a6d5110f986704c6e4da80a7a2263e0. Now it should return a map `{:id 2}` when querying.
